modules/EvseManager/BspStateMachine.cpp

```diff
@@ -0,0 +1,362 @@
+const std::string cpevent_to_string(CPEvent e) {
+    switch (e) {
+    case CPEvent::CarPluggedIn:
+        return "CarPluggedIn";
+    case CPEvent::CarRequestedPower:
+        return "CarRequestedPower";
+    case CPEvent::PowerOn:
+        return "PowerOn";
+    case CPEvent::PowerOff:
+        return "PowerOff";
+    case CPEvent::CarRequestedStopPower:
+        return "CarRequestedStopPower";
+    case CPEvent::CarUnplugged:
+        return "CarUnplugged";
+    case CPEvent::ErrorE:
+        return "ErrorE";
+    case CPEvent::ErrorDF:
+        return "ErrorDF";
+    case CPEvent::ErrorRelais:
+        return "ErrorRelais";
+    case CPEvent::ErrorVentilationNotAvailable:
+        return "ErrorVentilationNotAvailable";
+    case CPEvent::ErrorOverCurrent:
+        return "ErrorOverCurrent";
+    case CPEvent::ErrorOverVoltage:
+        return "ErrorOverVoltage";
+    case CPEvent::ErrorUnderVoltage:
+        return "ErrorUnderVoltage";
+    case CPEvent::ErrorMotorLock:
+        return "ErrorMotorLock";
+    case CPEvent::ErrorOverTemperature:
+        return "ErrorOverTemperature";
+    case CPEvent::ErrorBrownOut:
+        return "ErrorBrownOut";
+    case CPEvent::ErrorCablePP:
+        return "ErrorCablePP";
+    case CPEvent::ErrorEnergyManagement:
+        return "ErrorEnergyManagement";
+    case CPEvent::ErrorNeutralPEN:
+        return "ErrorNeutralPEN";
+    case CPEvent::ErrorCpDriver:
+        return "ErrorCpDriver";
+    case CPEvent::EFtoBCD:
+        return "EFtoBCD";
+    case CPEvent::BCDtoEF:
+        return "BCDtoEF";
+    case CPEvent::PermanentFault:
+        return "PermanentFault";
+    case CPEvent::EvseReplugStarted:
+        return "EvseReplugStarted";
+    case CPEvent::EvseReplugFinished:
+        return "EvseReplugFinished";
+    }
+
+    throw std::out_of_range("No known string conversion for provided enum of type CPEvent");
+}
+
+BspStateMachine::BspStateMachine(const std::unique_ptr<evse_board_supportIntf>& r_bsp) : r_bsp(r_bsp) {
+    // start state machine thread
+    main_thread_handle = std::thread(&BspStateMachine::main_thread, this);
+}
+
+void BspStateMachine::main_thread() {
+    while (true) {
+
+        if (main_thread_handle.shouldExit()) {
+            break;
+        }
+
+        std::this_thread::sleep_for(STATEMACHINE_UPDATE_RATE);
+
+        {
+            std::lock_guard l(state_mutex);
+            state_machine();
+        }
+    }
+}
+
+void BspStateMachine::process_bsp_event(types::board_support_common::BspEvent bsp_event) {
+
+    auto event_or_error = from_bsp_event(bsp_event.event);
+
+    // If it was a CP change: feed state machine, else forward error
+    if (std::holds_alternative<RawCPState>(event_or_error)) {
+        std::lock_guard l(state_mutex);
+        cp_state = std::get<RawCPState>(event_or_error);
+        state_machine();
+    } else {
+        event(std::get<CPEvent>(event_or_error));
+    }
+}
+
+void BspStateMachine::state_machine() {
```

Comment:
Disabled state: added. DF: Currently this is treated as a generic error (similar to e.g. over current error or generic CP error where no known CP state exists). This will put the Charger state machine into fault handling states which then will decide what to do

Reply:
They will be forwarded to charger in process_bsp_event() line 166:

    } else {
        event(std::get<CPEvent>(event_or_error));
    }